### PR TITLE
Test: Added unit tests for Plugins (5) - c8s65gxjrsjgb3brm8im6h9c74c

### PR DESCRIPTION
**Pull request by Hanan-Nawaz**
**Description**

- Added unit tests for:

  1. Reverse Osmosis Plugin
  2. Solar Concentrator Plugin
  3. Stored Power Electrolysis Plugin
  4. Variable Operating Cost Plugin

**Motivation / Background**

- Makes tests more readable, maintainable, and flexible.
- Ensures correct DCF and proper calculations.
- Ensures Plugins works accurately, when changes are made to them.

**Type of Change**

- [x] Test addition

**How Has This Been Tested?**

- All tests pass with uv run pytest -v.

**Checklist**

- [x]  Tests added and passing
- [x]  Comments added in critical sections
- [x]  Implemented all changes requested in last testing PR.

**Screenshot**
image

### CODE REVIEW

File: src/tests/plugins/reverse_osmosis_plugin_test.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        }
+        
+
+@pytest.mark.parametrize(
+    "case",
+    [
+        {
+            "input": {
+                "plant_output_per_year": np.array([100000.0]),      
+                "construction_time": 0,
```
Comment (axfache) [inline src/tests/plugins/reverse_osmosis_plugin_test.py]: I think constructio_time should be strictly positive

Comment (Hanan-Nawaz) [reply src/tests/plugins/reverse_osmosis_plugin_test.py]: Ok, I will update this. I thought, its only for that plugin.

File: src/tests/plugins/stored_power_electrolysis_plugin_test.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    [
+        {
+            "input": {
+                "fraction_stored_power_for_electrolysis": 0.8,
+                "electrolyzer_nominal_power_kW": 500.0,
+                "electrolyzer_power_increase_per_year": 0.02,
+                "electrolyzer_min_capacity": 0.1,
+                "electrolyzer_conversion_efficiency": 0.05,
+                "electrolyzer_replacement_time_h": 4000.0,
+                "electrolyzer_yearly_H2_production_kg": np.array([0.0, 0.0]),
```
Comment (axfache) [inline src/tests/plugins/stored_power_electrolysis_plugin_test.py]: The first value is 0 (construction year), the second value is expected to be positive

File: src/tests/plugins/stored_power_electrolysis_plugin_test.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                "electrolyzer_yearly_operation_hours": np.array(
+                    [[2026, 0, 0], [2027, 0, 0]]
```
Comment (axfache) [inline src/tests/plugins/stored_power_electrolysis_plugin_test.py]: same remark as before: testing the corner case is useful, but the typical case is not 0

Comment (Hanan-Nawaz) [reply src/tests/plugins/stored_power_electrolysis_plugin_test.py]: Implented.

File: src/tests/plugins/stored_power_electrolysis_plugin_test.py (hunk 2 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/stored_power_electrolysis_plugin_test.py]: Implented

File: src/tests/plugins/reverse_osmosis_plugin_test.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        {
+            "input": {
+                "plant_output_per_year": np.array([1800.0, 1900.0]),      
+                "construction_time": 1,                   
+                "power_demand_kWh_per_m3": 10.0,          
+                "avg_daily_hours": 24.0,                  
+                "recovery_rate": 0.1,                    
+            },
+            "expected": {
+                "electricity_demand_kWh": np.array([1702.977583544284]),
```
Comment (jschneidewind) [inline src/tests/plugins/reverse_osmosis_plugin_test.py]: The electricity_demand_kWh should be an array with two entries (since plant_output_per_year has two entries). I think the problem here is that plant_output_per_year should look like this : [0, 1800, 1900]. The first entry should be zero, since the construction time is one year. Please check this by running the PV+E reference file and have a look at the "Output per year" that is passed to the reverse osmosis plugin to check that this is the case and then adjust the "plant_output_per_year" accordingly.

File: src/tests/plugins/variable_operating_cost_plugin_test.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        plugin.utilities, 
+        expected["utilities"], 
+        rtol=tolerance
+    )
+    
+    np.testing.assert_allclose(
+        plugin.other, 
+        expected["other"], 
+        rtol=tolerance
+    )
```
Comment (jschneidewind) [inline src/tests/plugins/variable_operating_cost_plugin_test.py:87]: The solar thermal plugin is missing from the test suite, if I am not mistaken?

File: src/tests/plugins/stored_power_electrolysis_plugin_test.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                "Conversion efficiency (kg H2/kWh)": {
+                    "Value": electrolyzer_conversion_efficiency
+                },
+                "Replacement time (h)": {"Value": electrolyzer_replacement_time_h},
+                "H2 Production (yearly, kg)": {
+                    "Value": electrolyzer_yearly_H2_production_kg,
+                    "Processed": "Yes"
+                },
+                "Yearly Operation Data": {
+                    "Value": electrolyzer_yearly_operation_hours,
```
Comment (jschneidewind) [inline src/tests/plugins/stored_power_electrolysis_plugin_test.py]: The variable here should not be named "electrolyzer_yearly_operation_hours", since "yearly operation data" is not referring to the yearly operation hours,

**Review by jschneidewind — CHANGES_REQUESTED** [review]

File: src/tests/plugins/stored_power_electrolysis_plugin_test.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            "input": {
+                "fraction_stored_power_for_electrolysis": 0.8,
+                "electrolyzer_nominal_power_kW": 500.0,
+                "electrolyzer_power_increase_per_year": 0.02,
+                "electrolyzer_min_capacity": 0.1,
+                "electrolyzer_conversion_efficiency": 0.05,
+                "electrolyzer_replacement_time_h": 4000.0,
+                "electrolyzer_yearly_H2_production_kg": np.array([0.0, 2.0]),
+                "electrolyzer_yearly_operation_hours": np.array(
+                    [[2026, 3.0, 0], [2027, 4.0, 0]]
```
Comment (jschneidewind) [inline src/tests/plugins/stored_power_electrolysis_plugin_test.py]: For the yearly data (currently named "electrolyzer_yearly_operation_hours", the format is (year, H2 produced, electrolyzer capacity), hence the entries at index 1 should be identical to the values in "electrolyzer_yearly_H2_production_kg" and the entry at index 2 should be some value larger than 0 (at least for one of the entries)

File: src/tests/plugins/stored_power_electrolysis_plugin_test.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                ),
+                "stored_power_daily_kWh": {
+                    2026: np.array([1000.0, 1200.0, 0]),
+                    2027: np.array([900.0, 0, 700.0]),
+                },
+            },
+            "expected": {
+                "power_consumption_kWh": np.array([1760.0, 1280.0]),
+                "replacement_frequency": 2.0,
+                "new_h2_production_kg": np.array([3.3153904527033626e-16, 2.0000000000000004]),
```
Comment (jschneidewind) [inline src/tests/plugins/stored_power_electrolysis_plugin_test.py]: The new_h2_production_kg does not make sense (a value that is essentially 0 for 2026 and only 2 kg for 2027). Please look into what is happening here, we would expect roughly ~40 kg H2 production in both years (a bit less in 2027). The issue might be due to line 96 in the plugin. Please check what the plugin calculates for "additional_h2_production_kg" (in the plugin itself) and if this lines up with the expected values.

File: src/tests/plugins/variable_operating_cost_plugin_test.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+          
+
+@pytest.mark.parametrize(
+    "case",
+    [
+        {
+            "input": {
+                "plant_output_per_year": 100_000.0,  
+                "utilities": {
+                    "Electricity": {"Cost": 0.05, "Usage": 50.0},
```
Comment (jschneidewind) [inline src/tests/plugins/variable_operating_cost_plugin_test.py]: At least one of the utilities should have a conversion factor supplied

File: src/tests/plugins/variable_operating_cost_plugin_test.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+                "plant_output_per_year": 100_000.0,  
+                "utilities": {
+                    "Electricity": {"Cost": 0.05, "Usage": 50.0}, 
+                    "Water": {"Cost": 0.01, "Usage": 10.0}          
+                },
+                "other_variable_costs": {
+                    "Maintenance": 1000.0,
+                    "Chemicals": 500.0
+                },
+                "inflation_correction": 1.0,
```
Comment (jschneidewind) [inline src/tests/plugins/variable_operating_cost_plugin_test.py]: The inflation values should be some value larger than 1 (e.g. 1.1 and 1.2)

**Review by jschneidewind — CHANGES_REQUESTED** [review]

File: src/tests/plugins/reverse_osmosis_plugin_test.py (hunk 1 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/reverse_osmosis_plugin_test.py]: Implemented the change.

```python

"construction_time": 1

```

----------


**See line 38. (Reverse osmosis plugin)**

File: src/tests/plugins/stored_power_electrolysis_plugin_test.py (hunk 2 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/stored_power_electrolysis_plugin_test.py]: Implemented the change.

```python

"electrolyzer_yearly_H2_production_kg": np.array([1.0, 8.0]),

```

---------


**See line 69. (stored power electrolysis plugin)**

File: src/tests/plugins/stored_power_electrolysis_plugin_test.py (hunk 3 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/stored_power_electrolysis_plugin_test.py]: implemented the change.

```python

                "yearly_operation_hours": np.array([
                    [2026, 1.0, 10.0],
                    [2027, 8.0, 80.0]
                ]),

```

--------

**See line 71.  (stored power electrolysis plugin)**

File: src/tests/plugins/reverse_osmosis_plugin_test.py (hunk 4 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/reverse_osmosis_plugin_test.py]: Implemented the change. 


```python

[
        {
            "input": {
                "plant_output_per_year": np.array([0.0, 1800.0, 1900.0]),      
                "construction_time": 1,                   
                "power_demand_kWh_per_m3": 10.0,          
                "avg_daily_hours": 24.0,                  
                "recovery_rate": 0.1,                    
            },
            "expected": {
                "electricity_demand_kWh": np.array([1613.3471844103742, 1702.977583544284]),
                "max_capacity_m3_per_hour": 0.019440383373793193
            },
        }
    ]
``` 

- updated the plant_output_year as described
- updated the electricity_demand_kwh


-----------


**See Lines from 34 to 48. (Reverse osmosis plugin)**


----


which refernce file? we have PV_E in multiple places.

File: src/tests/plugins/variable_operating_cost_plugin_test.py (hunk 5 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/variable_operating_cost_plugin_test.py:87]: It is already added in the first PR.


See: #11

File: src/tests/plugins/stored_power_electrolysis_plugin_test.py (hunk 6 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/stored_power_electrolysis_plugin_test.py]: done.

File: src/tests/plugins/stored_power_electrolysis_plugin_test.py (hunk 7 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/stored_power_electrolysis_plugin_test.py]: done.

File: src/tests/plugins/stored_power_electrolysis_plugin_test.py (hunk 8 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/stored_power_electrolysis_plugin_test.py]: checked. same as above.

Calculated additional_h2_production_kg: [3.31539045e-16, 2.36391476e-16]

Calculated new_h2_production_kg: [1. 8.]

File: src/tests/plugins/variable_operating_cost_plugin_test.py (hunk 9 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/variable_operating_cost_plugin_test.py]: done.

File: src/tests/plugins/variable_operating_cost_plugin_test.py (hunk 10 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/variable_operating_cost_plugin_test.py]: done.

**Review by jschneidewind — APPROVED** [review]
Looks good, thanks!
